Bug Fixed GPTNeoX Flax supports

HeegyuKim opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->
Fixes #22950:
- previous FlaxGPTNeoX support PR #22950 contains a error in the cached generation process. I resolved it.
- And I inserted a test code from #24002 
- There were 7 failures in the course of testing, I'm not sure why. It doesn't seem like a very fatal issue, but I'd appreciate it if you could check it out.
- The output with the pytorch model is very similar and the model works fine.
- k/v cache is already implemented.

@sanchit-gandhi

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

HeegyuKim committed:
add code

HeegyuKim committed:
implementa flax gptneox model & tests

HeegyuKim committed:
style change

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# source code from https://github.com/huggingface/transformers/pull/22950
```
```suggestion
""" Flax GPT NeoX model."""
```

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+
+
+# positional embedding
+class RotaryEmbeddingNP(linen.Module):
```
Not sure about the naming here - shouldn't it match the PyTorch model name?
```suggestion
class FlaxGPTNeoXRotaryEmbedding(linen.Module):
```

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+# positional embedding
+class RotaryEmbeddingNP(linen.Module):
+    dim: int
+    max_seq_len_cached: int
```
Argument name diverges from PyTorch implementation: https://github.com/huggingface/transformers/blob/3a05e010e0c7e8abd3e5357dd4e89e28cc69003e/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L298

Is there a reason for this?

HeegyuKim replied:
No, I'll unify the name

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        self.inv_freq = 1.0 / (self.base ** (jnp.arange(0, self.dim, 2).astype(jnp.float32) / self.dim))  # dim
+
+    def __call__(self, x=None, seq_len=None):
+        t = jnp.arange(self.max_seq_len_cached, dtype=self.inv_freq.dtype)
```
Hemm we're not actually using any cached embeddings here - we compute them on-the-fly each time the embedding layer is called. I think we can copy the logic used in the PyTorch code and cache the embeddings if the `seq_len` of the input is less than our `self.max_seq_len_cached`. Only if `seq_len > self.max_seq_len_cached` do we re-compute the embeddings

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        self.causal_mask = make_causal_mask(jnp.ones((1, config.max_position_embeddings), dtype="bool"), dtype="bool")
+
+    @linen.compact
+    def _concatenate_to_cache(self, key, value, query, attention_mask):
```
Let's try and use `# Copied from...` everywhere we re-use existing modelling code

```suggestion
# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoSelfAttention._concatenate_to_cache
def _concatenate_to_cache(self, key, value, query, attention_mask):
```

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+    config: GPTNeoXConfig
+    dtype: jnp.dtype = jnp.float32
+    causal: bool = True
+    is_cross_attention: bool = False
```
Argument un-used
```suggestion

```

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+            attention_mask = combine_masks(pad_mask, attention_mask)
+        return key, value, attention_mask
+
+    def _split_heads(self, hidden_states):
```
Method not used - can remove or update

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        qkv = self.query_key_value(hidden_states)
+        batch, seq_len, _ = qkv.shape
+
+        qkv = qkv.reshape([batch, seq_len, self.num_attention_heads, 3, self.head_size])
```
Let's either move this to `_split_heads` or remove the method entirely

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        else:
+            causal_mask = self.causal_mask[:, :, :query_length, :key_length]
+
+        batch_size = hidden_states.shape[0]
```
Already determined batch size above (L225)

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        self.dense_4h_to_h = nn.Dense(embed_dim, dtype=self.dtype, kernel_init=kernel_init)
+
+        self.act = ACT2FN[self.config.hidden_act]
+        # self.dropout = nn.Dropout(rate=self.config.resid_pdrop)
```
Let's remove any un-used comments

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        self.act = ACT2FN[self.config.hidden_act]
+        # self.dropout = nn.Dropout(rate=self.config.resid_pdrop)
+
+    def __call__(self, hidden_states, deterministic: bool = True):
```
No need for `deterministic` arg if we don't use dropout

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        hidden_states = self.dense_h_to_4h(hidden_states)
+        hidden_states = self.act(hidden_states)
+        hidden_states = self.dense_4h_to_h(hidden_states)
+        # hidden_states = self.dropout(hidden_states, deterministic=deterministic)
```
Can remove

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+    config: GPTNeoXConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
```
(nit) can we organise this method so that it follows the order in which the modules are used?

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.post_attention_layernorm = nn.LayerNorm(epsilon=self.config.layer_norm_eps)
```
Missing dtype arg
```suggestion
        self.post_attention_layernorm = nn.LayerNorm(epsilon=self.config.layer_norm_eps, dtype=self.dtype)
```

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        return (hidden_states,) + attn_outputs[1:]
+
+
+class FlaxGPTNeoXPreTrainedModel(FlaxPreTrainedModel):
```
We can probably add a `# Copied from...` here and copy the entire class. If there are individual methods that differ, let's just use `# Copied from...` on the individual methods that are the same, and omit it for the new ones
```suggestion
# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel with GPTNeo->GPTNeoX
class FlaxGPTNeoXPreTrainedModel(FlaxPreTrainedModel):
```

sanchit-gandhi replied:
Just checked and the `__call__` method is different, so let's use `# Copied from...` on all but this method

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        return outputs
+
+
+class FlaxGPTNeoXBlockCollection(nn.Module):
```
Can also be copied

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+        return_dict: bool = True,
+    ):
+        batch_size, seq_length = input_ids.shape
+        if attention_mask is None:
```
This shouldn't go here -> should be done in `FlaxGPTNeoXPreTrainedModel.__call__`

HeegyuKim replied:
Thank you, both attention_mask and position_ids are already handled in `FlaxGPTNeoXPreTrainedModel.__call__` so that I'll remove it

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,722 @@
+            attention_mask = jnp.ones_like(input_ids)
+
+        if position_ids is None:
+            position_ids = jnp.broadcast_to(
```
This shouldn't go here -> should be done in `FlaxGPTNeoXPreTrainedModel.__call__`

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,336 @@
+    def __init__(
+        self,
+        parent,
+        batch_size=13,
```
Let's use bs=2
```suggestion
        batch_size=2,
```

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,336 @@
+        use_labels=True,
+        vocab_size=99,
+        hidden_size=32,
+        num_hidden_layers=5,
```
Let's use 2 hidden layers and 2 attention heads to speed up the CI!

sanchit-gandhi reviewed (commented):
Looking super clean already @HeegyuKim! Mainly just some very minor comments from me - the overall design is great. Once these are addressed and the tests pass we can get it ready for merge 🚀

HeegyuKim committed:
자잘한 변경사항 처리(주석, 안쓰이는 기능들, 변수명 등)

HeegyuKim committed:
resolve test error

HeegyuKim committed:
definition order changes

HeegyuKim committed:
use RoPE for GPT-j

HeegyuKim committed:
"copy from" comments

HeegyuKim commented:
I suffering from test issue. Can you help me? @sanchit-gandhi 

summary
- In pytorch, my GPTNeoX test failed to both test_equivalence_flax_to_pt and test_equivalence_pt_to_flax tests
- But in flax, GPTNeoX doesn't failed because my test code overrides it for not using `check_pt_flax_outputs`
- It's same to GPT Neo, Flax GPTNeo don't use `check_pt_flax_outputs` but Pytorch GPTNeo use it
- However, GPT Neo test in pytorch do not fail.
- Flax GPTNeo fails if it uses `check_pt_flax_outputs`

I don't think this is a problem with my model implementation. I wonder why pytorch's test fails.


This PR failed two tests below
```
FAILED tests/models/gpt_neox/test_modeling_gpt_neox.py::GPTNeoXModelTest::test_equivalence_flax_to_pt - AssertionError: 1.0483556 not less than or equal to 1e-05 : outputs.last_hidden_state: Difference between PyTorch and Flax is 1.0483555793762207 (>= 1e-05).
FAILED tests/models/gpt_neox/test_modeling_gpt_neox.py::GPTNeoXModelTest::test_equivalence_pt_to_flax - AssertionError: 1.8777691 not less than or equal to 1e-05 : outputs.last_hidden_state: Difference between PyTorch and Flax is 1.877769112586975 (>= 1e-05).
```
But two flax tests in tests/models/gpt_neox/test_modeling_flax_gptneox.py are fine.

the test code which was copied from #24002 override both test_equivalence_pt_to_flax and test_equivalence_flax_to_pt methods with this comment.
```
    # overwrite from common since `attention_mask` in combination
    # with `causal_mask` behaves slighly differently
```

and they use below assert code
```
# test_modeling_flax_gptneox.py line 267
for fx_output, pt_output in zip(fx_outputs, pt_outputs):
    self.assert_almost_equals(fx_output[:, -1], pt_output[:, -1].numpy(), 4e-2)
```

instead of `self.check_pt_flax_outputs(fx_outputs, pt_outputs, model_class)` in test_equivalence_pt_to_flax method in test_modeling_common.py

This overrides are equal to `tests/models/gpt_neo/test_modeling_flax_gptneo.py` but GPTNeo doesn't fail to pytorch test. I don't know what is different and

HeegyuKim committed:
remove random attention masking

HeegyuKim committed:
black formatting

liutianlin0121 commented:
Hi! Thanks @HeegyuKim for the PR. I am wondering if there is any update on this? It would be really cool if we could use GPTNeoXForCausalLM in flax!

HeegyuKim commented:
Hello @liutianlin0121 I'm trying to solve the problem whenever I have time. However, even if causal masking is applied, the error in the model output is still larger than 1e-5. The current error is around 0.02-0.03. I'm going to try again this weekend.

Even though there are errors, the model works better than expected. I trained several models with this code.

- [heegyu/WizardVicuna-pythia-1.4b-deduped](https://huggingface.co/heegyu/WizardVicuna-pythia-1.4b-deduped)
- [heegyu/RedTulu-Uncensored-3B-0719](https://huggingface.co/heegyu/RedTulu-Uncensored-3B-0719)
- [heegyu/polyglot-ko-5.8b-chat](https://huggingface.co/heegyu/polyglot-ko-5.8b-chat)

I want to contribute to huggingface but it's not as easy as I thought.

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,339 @@
+
+        input_mask = None
+        if self.use_input_mask:
+            input_mask = torch.tril(torch.ones(self.batch_size, self.seq_length)).numpy()
```
This can probably be done entirely in `numpy`, i.e. 
```python
            input_mask = np.tril(np.ones((self.batch_size, self.seq_length)))
```
=> this way, you don't have to import `torch` in this Flax test file

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,339 @@
+    # overwrite from common since `attention_mask` in combination
+    # with `causal_mask` behaves slighly differently
+    @is_pt_flax_cross_test
+    def test_equivalence_pt_to_flax(self):
```
We ideally shouldn't have to override this test for the final PR, since it should pass with a causal attention mask

sanchit-gandhi reviewed (commented):
It looks like you're making solid progress @HeegyuKim! Nice work! The most recent CI run is reporting that the difference in the Flax and PyTorch model outputs is > 1 (see [CI Output](https://app.circleci.com/pipelines/github/huggingface/transformers/73102/workflows/a67744dc-ced9-4587-b905-d79622bc9e70/jobs/924001)). This suggests to me that there is a divergence between the Flax and PyTorch models. Generally, for any model less than 1B params, we should be able to get equivalence to within 1e-5 between Flax and PyTorch. It's quite likely that you won't get this equivalence running the matmuls in bfloat16 on TPU. But you should be able to running the matmuls in float32, see https://github.com/huggingface/transformers/issues/15754 and https://github.com/google/jax/issues/10413#issue-1212211265 for details

Here's a script that I used previously for checking PT / Flax equivalence for BLOOM: https://github.com/sanchit-gandhi/codesnippets/blob/main/check_flax_bloom_jit_small_testing.ipynb You can ignore the bits about JIT'ing the forward pass for the time being. You can also uncomment the check to run it on CPU to force the highest precision, or use the decorator as provided

If we don't get 1e-5 precision, it's usually an indicator that we have a divergence in our model. Here, going through layer-by-layer and checking the hidden-states might be required to pinpoint it. Once you have this equivalence, it's almost guaranteed that the CI will report a difference of less than 1e-5, since it runs on CPU.

Let me know if you have any questions / queries about finishing this PR. You've done a great job and I'd be more than happy to assist you in seeing this to completion!

HeegyuKim committed:
tril attention masking for pt, use numpy for flax

HeegyuKim committed:
replace RoPE from EasyDeL

HeegyuKim committed:
code consistency, remove gptneox flax test override

HeegyuKim committed:
remove comment, GPTNeoX caching, renaming

HeegyuKim committed:
black formatiing, code copy check, enable flax in index.md

HeegyuKim committed:
ruff  fix

HeegyuKim commented:
Ohhhhh I finally pass the equivalence issue! 🎉🎉

- I use FlaxGPTNeoXRotaryEmbedding class for RoPE and implement caching.
- I remove overrides in tests/models/gpt_neox/test_modeling_flax_gpt_neox.py and it works!
 
But there are CI failures...
- [check_repository_consistency](https://app.circleci.com/pipelines/github/huggingface/transformers/73610/workflows/4216a793-3879-4e27-85de-9aefe7e75998/jobs/931100): I copied it identically, but why does this message appear?
- [tests_pr_documentation_tests](https://app.circleci.com/pipelines/github/huggingface/transformers/73610/workflows/00dfecbe-027f-41a7-93ff-7161660486f3/jobs/931111): Huggingface's GPT-NeoX 20B model does not have a flax weight. How should we solve this problem?

@sanchit-gandhi

sanchit-gandhi commented:
Well done @HeegyuKim, that's excellent news! Regarding the two failing tests:
* You can run `make fix-copies` to update the modelling code with any copied functions? The linter will copy all the code so that it is one-for-one the same
* Could you open a pull request on the Hugging Face Hub to add the Flax weights to [EleutherAI/gpt-neox-20b](https://huggingface.co/EleutherAI/gpt-neox-20b)? You can convert first load the PyTorch weights into Flax:
```python
from transformers import FlaxAutoModelForCausalLM

model = FlaxAutoModelForCausalLM.from_pretrained("EleutherAI/gpt-neox-20b", from_pt=True)
```

And then push the converted Flax weights to the Hub:
```python
model.push_to_hub("EleutherAI/gpt-neox-20b", create_pr=True)
```

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,732 @@
+
+class FlaxGPTNeoXMLP(nn.Module):
+    config: GPTNeoXConfig
+    intermediate_size: int
```
Do we need to pass the `intermediate_size` argument as an attribute? We can just get it from `config.intermediate_size` right?

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,732 @@
+        output_hidden_states: bool = False,
+        return_dict: bool = True,
+    ):
+        input_embeds = self.embed_in(input_ids.astype("i4"))
```
Strictly speaking there shouldn't be a need to convert `input_ids` to type "i4" - this is done in the `__call__` method of `FlaxGPTNeoXPreTrainedModel`: https://github.com/huggingface/transformers/blob/ef0c6d9818dbf563ec392472d9669f6b43bda6ab/src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py#L496

HeegyuKim replied:
If we remove this conversion, it occurs errors in generation tests.
```
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_beam_search_generate - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_beam_search_generate_attn_mask - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_beam_search_generate_logits_warper - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_beam_search_generate_num_return_sequences - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_greedy_generate - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_greedy_generate_attn_mask - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_greedy_generate_logits_warper - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_greedy_generate_pt_fx - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_sample_generate - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_sample_generate_attn_mask - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_sample_generate_logits_warper - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_use_cache_forward - ValueError: Input type must be an integer or unsigned integer.
FAILED tests/models/gpt_neox/test_modeling_flax_gpt_neox.py::FlaxGPTNeoXModelTest::test_use_cache_forward_with_attn_mask - ValueError: Input type must be an integer or unsigned integer.
```

GPTNeoModule converts input_ids too and has same problems.
https://github.com/huggingface/transformers/blob/ef0c6d9818dbf563ec392472d9669f6b43bda6ab/src/transformers/models/gpt_neo/modeling_flax_gpt_neo.py#L549


I'm looking for the cause. Please let me know if you know the right solution

sanchit-gandhi replied:
Okay cool let's leave it as is for consistency

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,732 @@
+    ):
+        input_embeds = self.embed_in(input_ids.astype("i4"))
+
+        hidden_states = input_embeds
```
```suggestion
        hidden_states = self.emb_dropout(input_embeds, deterministic=deterministic)
```

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,231 @@
+# Copyright 2021 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2023 The HuggingFace Team. All rights reserved.
```

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,231 @@
+        type_vocab_size=16,
+        type_sequence_label_size=2,
+        initializer_range=0.02,
+        num_labels=3,
```
The last three init arguments are not used in the config - let's remove them?

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,231 @@
+
+        input_mask = None
+        if self.use_input_mask:
+            input_mask = np.tril(np.ones((self.batch_size, self.seq_length)))
```
Note to reviewer: we use a causal attention mask in the Flax generation tests. This is required to get sensible outputs from the Flax model (which is typical behaviour).

sanchit-gandhi commented on tests/models/gpt_neox/test_modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,231 @@
+            attention_probs_dropout_prob=self.attention_probs_dropout_prob,
+            max_position_embeddings=self.max_position_embeddings,
+            type_vocab_size=self.type_vocab_size,
+            is_decoder=False,
```
Should be `True` no?

sanchit-gandhi reviewed (commented):
Looks in great shape @HeegyuKim! And nice job on getting equivalence with PyTorch! Left a few suggestions below, mainly just small re-factoring to get it ready for merge. Feel free to ping me as soon as you're ready for a final look - think it should be pretty fast to get it merged from here

HeegyuKim committed:
modify pytest from review

HeegyuKim committed:
resolve simple modeling review

HeegyuKim committed:
RoPE scaling copied but SetAttributeFrozenModuleError exists

HeegyuKim commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,732 @@
+
+        self.causal_mask = make_causal_mask(jnp.ones((1, config.max_position_embeddings), dtype="bool"), dtype="bool")
+
+        self.rotary_emb = FlaxGPTNeoXRotaryEmbedding(
```
I copied RoPE scaling code and test code. But in [test CI](https://app.circleci.com/pipelines/github/huggingface/transformers/74067/workflows/e61221ea-2282-4e69-ae4c-1fa4b0e28f30/jobs/937874), the frozen flax model raises SetAttributeFrozenModuleError.

For the cached RoPE embedding, I think I should use the variable. I'm trying to implement it, and I think it'd be nice if you had an appropriate reference or suggestion. @sanchit-gandhi

sanchit-gandhi replied:
Ah we can only set attributes in Flax in the `setup` method. After that, the module gets frozen, meaning we can't add new attributes or update existing ones. So probably the cached embedding isn't going to work - what we can instead do is always initialise the embeddings to max length (`config.max_position_embeddings`), and then slice the first `N` embeddings as required each time. This way, we don't ever need to re-compute or update the embeddings, since we always have the max embedding length we require stored in the `setup`

HeegyuKim committed:
resolve rope scaling test issue

HeegyuKim committed:
add gelu_fast

HeegyuKim committed:
modify "copied from" comment

HeegyuKim committed:
insert nn.compact

HeegyuKim committed:
jnp.DeviceArray -> jax.Array

HeegyuKim committed:
change order of is_pt_flax_cross_test

HeegyuKim committed:
modify text generation tests

HeegyuKim commented:
I think we're almost at the end of our work but there are small issues.

### Suddenly wav2vec2 test fails??
- Suddenly wav2vec model tests are failed ([tests_torch CI link](https://app.circleci.com/pipelines/github/huggingface/transformers/74617/workflows/f61c5c58-d39a-4394-8ebc-8b551d74da3b/jobs/945928)). It seems to have something to do with gelu_fast that I added. I don't think wav2vec2 uses gelu_fast, but I don't know why.
- GPT-NeoX-20B uses gelu_fast activation. I converted pytorch implemented to flax version in [src/transformers/modeling_flax_utils.py](src/transformers/modeling_flax_utils.py) and added gelu_fast. 

### Flax weights
- I opened the PR of GPT-NeoX 20B Flax weights - https://huggingface.co/EleutherAI/gpt-neox-20b/discussions/24
- the Flax tests use pythia-410m models but there is no flax weights. So I use `from_pt=True` is it Ok?

### `Copied from` issue
- `make fix-copies` changes every `GPTNeoXBlahBlah` -> `GPTNeoBlahBlah` (config, comments) even there is a `Copied from ... with GPTNeo->GPTNeoX` mark. https://github.com/huggingface/transformers/pull/25334/commits/a670443b2e4fdd88262c93eb98a9dc0330c64812#r1287443977 
- So I moved `copied from` comments to the exactly same method.

@sanchit-gandhi

sanchit-gandhi commented:
Hey @HeegyuKim! Nice job on iterating here! Answering your questions in-line below:
1. For the failing Wav2Vec2 issues, you can rebase onto the main branch of `transformers`, where the tests have been fixed:
```
git fetch upstream
git rebase upstream main
git push -f origin main
```
Note that it's important you force push (`-f` flag) after a rebase to preserve the correct commit history of this PR!
2. Thanks for converting the PyTorch GELU Fast activation to JAX - this looks great!
3. Thanks also for pushing the Flax weights - we can merge them once this PR is approved by the next reviewer and prior to merging this PR. Using `from_pt=True` is ok for the Flax tests - just make sure you decorate the tests with `@is_pt_flax_cross_test` since we need both PyTorch and Flax when we load from pre-trained with `from_pt=True`
4. The issue you had before was that there was an extra space after the start and end of the right arrow: `GPTNeo -> GPTNeoX` should be `GPTNeo->GPTNeoX`. Can you try adding this copied from before `FlaxGPTNeoXPreTrainedModel`? This should allow you to copy the entire module:
# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel with GPTNeo->GPTNeoX, "transformer"->"gpt_neox", GPT_NEO->GPT_NEO_X
```

HeegyuKim committed:
Revert "modify "copied from" comment"

This reverts commit a670443b2e4fdd88262c93eb98a9dc0330c64812.

HeegyuKim committed:
pass make fix-copies

HeegyuKim committed:
add pt_flax cross test annotation

HeegyuKim committed:
black

HeegyuKim commented:
Finally documentation is left, how can I make a documentation for it? @sanchit-gandhi 
```
Exception: The following objects are in the public init so should be documented:
 - FlaxGPTNeoXForCausalLM
 - FlaxGPTNeoXModel
```

sanchit-gandhi commented:
You can do so with `make repo-consistency`!

HeegyuKim committed:
modify consistency issue

HeegyuKim commented:
I may passed necessary CI tests! @sanchit-gandhi

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,783 @@
+            return self.cos_cached, self.sin_cached
+
+    def _compute_cos_sin(self, seq_len):
+        t = jnp.arange(seq_len, dtype=self.inv_freq.dtype)
```
Note to reviewer: single-letter variables chosen to maintain equivalence with the PyTorch modelling code https://github.com/huggingface/transformers/blob/2c658b5a4282f2e824b4e23dc3bcda7ef27d5827/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L303

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,783 @@
+    return jnp.concatenate((-second_half, first_half), axis=-1)
+
+
+class FlaxGPTNeoXRotaryEmbedding(nn.Module):
```
The embedding classes are very nice! Ported to JAX while following closely the logic from PyTorch

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,783 @@
+        return unfreeze(init_variables["cache"])
+
+    @add_start_docstrings_to_model_forward(GPT_NEOX_INPUTS_DOCSTRING)
+    def __call__(
```
Can we not copy this entire method from GPT Neo as well? The code looks to be one-to-one the same now, it's just a comment which is different "`# if past_key_values are passed...`"

If we do so, then we can actually just copy the entire class from GPT Neo, which would make the copied from statements much simpler.

sanchit-gandhi commented on src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py:
```diff
@@ -0,0 +1,783 @@
+
+        hidden_states = outputs[0]
+
+        lm_logits = self.embed_out(hidden_states)
```
What about possibly tied word embeddings? https://github.com/huggingface/transformers/blob/2c658b5a4282f2e824b4e23dc3bcda7ef27d5827/src/transformers/models/gpt_neo/modeling_flax_gpt_neo.py#L635-L637

sanchit-gandhi reviewed (approved):
Very nice @HeegyuKim! Sorry about the delay with getting you another review. It's mainly small nits from me. Let's put in the request for a core maintainer to take a final look at this and get it merged!

ArthurZucker reviewed (approved):
Great work here! @HeegyuKim Thanks for adding the flax support 🤗

sanchit-gandhi commented:
The PR is almost finished! Would you like to make the last remaining changes @HeegyuKim such that we can get this one merged? Let us know if you have any questions, more than happy to help here

HeegyuKim commented:
Thank you for your comment! I'll check it this weekend

HeegyuKim commented:
Hello @sanchit-gandhi, I rebased this PR to main branch and pushed again.

There are two CI failures - First is a [documentation issue](https://app.circleci.com/pipelines/github/huggingface/transformers/82163/workflows/26395aa8-e716-4cc0-a97b-8ad4c7ffe89b/jobs/1056756). 
```
OSError: Can't load the model for 'EleutherAI/gpt-neox-20b'. If you were trying to load it from 'https://huggingface.co/models', make sure you don't have a local directory with the same name. Otherwise, make sure 'EleutherAI/gpt-neox-20b' is the correct path to a directory containing a file named flax_model.msgpack or pytorch_model.bin.
```
This problem can be fixed when this [GPT-NeoX model PR](https://huggingface.co/EleutherAI/gpt-neox-20b/discussions/24) is merged. Alternatively, we can add from_pt=True to the example. 

As for the [second issue](https://app.circleci.com/pipelines/github/huggingface/transformers/82163/workflows/216462aa-cd92-40d8-ab8e-3e4103bc29b5/jobs/1056753), I don't know why. I would appreciate it if you could tell me the cause and solution to this problem.
```
ValueError: The main __init__ has objects that are not present in transformers.utils.dummy_flax_objects.py. Run `make fix-copies` to fix this.
```

I ran `make fix-copies` but following error occurs.
```
> make fix-copies

python utils/check_copies.py --fix_and_overwrite
Traceback (most recent call last):
  File "/home/heegyu/transformers/utils/check_copies.py", line 1129, in <module>
    check_copies(args.fix_and_overwrite, args.file)
  File "/home/heegyu/transformers/utils/check_copies.py", line 778, in check_copies
    new_diffs = is_copy_consistent(filename, overwrite, buffer)
  File "/home/heegyu/transformers/utils/check_copies.py", line 736, in is_copy_consistent
    diff_index = check_codes_match(observed_code, theoretical_code)
  File "/home/heegyu/transformers/utils/check_copies.py", line 549, in check_codes_match
    theoretical_name = re_pattern.search(theoretical_code_header).groups()[0]
AttributeError: 'NoneType' object has no attribute 'groups'
```
